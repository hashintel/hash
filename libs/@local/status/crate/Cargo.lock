# This file is automatically @generated by Cargo.
# It is not intended for manual editing.
version = 3

[[package]]
name = "anyhow"
version = "1.0.66"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "216261ddc8289130e551ddcd5ce8a064710c0d064a4d2895c67151c92b5443f6"

[[package]]
name = "atty"
version = "0.2.14"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "d9b39be18770d11421cdb1b9947a45dd3f37e93092cbf377614828a319d5fee8"
dependencies = [
 "hermit-abi",
 "libc",
 "winapi",
]

[[package]]
name = "error-stack"
version = "0.2.4"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "859d224e04b2d93d974c08e375dac9b8d1a513846e44c6666450a57b1ed963f9"
dependencies = [
 "anyhow",
 "owo-colors",
 "rustc_version",
]

[[package]]
name = "hash-status"
version = "0.0.0"
dependencies = [
 "error-stack",
 "serde",
 "serde_json",
]

[[package]]
name = "hermit-abi"
version = "0.1.19"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "62b467343b94ba476dcb2500d242dadbb39557df889310ac77c5d99100aaac33"
dependencies = [
 "libc",
]

[[package]]
name = "is_ci"
version = "1.1.1"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "616cde7c720bb2bb5824a224687d8f77bfd38922027f01d825cd7453be5099fb"

[[package]]
name = "itoa"
version = "1.0.4"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "4217ad341ebadf8d8e724e264f13e593e0648f5b3e94b3896a5df283be015ecc"

[[package]]
name = "libc"
version = "0.2.138"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "db6d7e329c562c5dfab7a46a2afabc8b987ab9a4834c9d1ca04dc54c1546cef8"

[[package]]
name = "owo-colors"
version = "3.5.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "c1b04fb49957986fdce4d6ee7a65027d55d4b6d2265e5848bbb507b58ccfdb6f"
dependencies = [
 "supports-color",
]

[[package]]
name = "proc-macro2"
version = "1.0.47"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "5ea3d908b0e36316caf9e9e2c4625cdde190a7e6f440d794667ed17a1855e725"
dependencies = [
 "unicode-ident",
]

[[package]]
name = "quote"
version = "1.0.21"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "bbe448f377a7d6961e30f5955f9b8d106c3f5e449d493ee1b125c1d43c2b5179"
dependencies = [
 "proc-macro2",
]

[[package]]
name = "rustc_version"
version = "0.4.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "bfa0f585226d2e68097d4f95d113b15b83a82e819ab25717ec0590d9584ef366"
dependencies = [
 "semver",
]

[[package]]
name = "ryu"
version = "1.0.11"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "4501abdff3ae82a1c1b477a17252eb69cee9e66eb915c1abaa4f44d873df9f09"

[[package]]
name = "semver"
version = "1.0.14"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "e25dfac463d778e353db5be2449d1cce89bd6fd23c9f1ea21310ce6e5a1b29c4"

[[package]]
name = "serde"
version = "1.0.150"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "e326c9ec8042f1b5da33252c8a37e9ffbd2c9bef0155215b6e6c80c790e05f91"
dependencies = [
 "serde_derive",
]

[[package]]
name = "serde_derive"
version = "1.0.150"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "42a3df25b0713732468deadad63ab9da1f1fd75a48a15024b50363f128db627e"
dependencies = [
 "proc-macro2",
 "quote",
 "syn",
]

[[package]]
name = "serde_json"
version = "1.0.89"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "020ff22c755c2ed3f8cf162dbb41a7268d934702f3ed3631656ea597e08fc3db"
dependencies = [
 "itoa",
 "ryu",
 "serde",
]

[[package]]
name = "supports-color"
version = "1.3.1"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "8ba6faf2ca7ee42fdd458f4347ae0a9bd6bcc445ad7cb57ad82b383f18870d6f"
dependencies = [
 "atty",
 "is_ci",
]

[[package]]
name = "syn"
version = "1.0.105"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "60b9b43d45702de4c839cb9b51d9f529c5dd26a4aff255b42b1ebc03e88ee908"
dependencies = [
 "proc-macro2",
 "quote",
 "unicode-ident",
]

[[package]]
name = "unicode-ident"
version = "1.0.5"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "6ceab39d59e4c9499d4e5a8ee0e2735b891bb7308ac83dfb4e80cad195c9f6f3"

[[package]]
name = "winapi"
version = "0.3.9"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "5c839a674fcd7a98952e593242ea400abe93992746761e38641405d28b00f419"
dependencies = [
 "winapi-i686-pc-windows-gnu",
 "winapi-x86_64-pc-windows-gnu",
]

[[package]]
name = "winapi-i686-pc-windows-gnu"
version = "0.4.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "ac3b87c63620426dd9b991e5ce0329eff545bccbbb34f3be09ff6fb6ab51b7b6"

[[package]]
name = "winapi-x86_64-pc-windows-gnu"
version = "0.4.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "712e227841d057c1ee1cd2fb22fa7e5a5461ae8e48fa2ca79ec42cfc1931183f"
