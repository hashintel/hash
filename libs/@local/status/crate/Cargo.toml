[package]
name = "hash-status"
version = "0.0.0"
edition = "2021"
description = "The HASH Status and Error Model."

[dependencies]
error-stack = { version = "0.3.1" }
serde = { version = "1.0.164", features = ["derive"] }
serde_json = "1.0.99"
