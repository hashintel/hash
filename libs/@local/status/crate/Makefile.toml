extend = { path = "../../../../.github/scripts/rust/Makefile.toml" }

# This crate has no unsafe code
[tasks.miri]
disabled = true
