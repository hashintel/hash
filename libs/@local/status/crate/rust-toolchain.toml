[toolchain]
channel = "nightly-2023-08-07"
components = ['rustfmt', 'clippy', 'llvm-tools-preview']
