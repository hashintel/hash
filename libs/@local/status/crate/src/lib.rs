// Attribution: *Heavily* inspired by the Google Cloud API Error Model
//  https://cloud.google.com/apis/design/errors

//! Defines a logical status and error model that is suitable for different programming
//! environments, including REST APIs and RPC APIs.

mod status_code;

use std::fmt::Debug;

use serde::{Deserialize, Serialize};
pub use status_code::StatusCode;

/// The canonical shape of a response object describing the status of a request between services.
#[derive(Clone, Eq, Debug, PartialEq, Serialize, Deserialize)]
pub struct Status<D>
where
    D: Send + Sync + Debug,
{
    code: StatusCode,
    /// A developer-facing description of the status.
    ///
    /// Where possible, this should provide guiding advice for debugging and/or handling the error.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    message: Option<String>,
    contents: Vec<D>,
}

impl<D> Status<D>
where
    D: Send + Sync + Debug + Serialize + for<'de> Deserialize<'de>,
{
    #[must_use]
    pub fn new(code: StatusCode, message: Option<String>, contents: Vec<D>) -> Self {
        Self {
            code,
            message,
            contents,
        }
    }

    #[must_use]
    pub const fn code(&self) -> StatusCode {
        self.code
    }

    #[must_use]
    pub const fn message(&self) -> &Option<String> {
        &self.message
    }

    #[must_use]
    pub fn contents(&self) -> &[D] {
        &self.contents
    }
}
