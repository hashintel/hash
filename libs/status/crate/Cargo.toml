[package]
name = "hash-status"
version = "0.0.0"
edition = "2021"
description = "The HASH Status and Error Model."

[dependencies]
error-stack = { version = "0.2.4" }
serde = { version = "1.0.148", features = ["derive"] }
serde_json = "1.0.89"
