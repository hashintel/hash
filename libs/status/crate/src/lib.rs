//! Defines a logical status and error model that is suitable for different programming
//! environments, including REST APIs and RPC APIs.

mod status_code;

pub use status_code::StatusCode;
