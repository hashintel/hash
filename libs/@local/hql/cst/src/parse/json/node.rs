use error_stack::{Report, Result, ResultExt};
use hql_cst_lex::{Lexer, Location, SyntaxKind, SyntaxKindSet, Token, TokenKind};
use winnow::{combinator::alt, Parser};

use super::{
    node_object::NodeObjectParser,
    util::{ArrayParser, EofParser},
    WinnowError,
};
use crate::{
    arena::Arena,
    expr::{call::Call, path::parse_path, signature::parse_signature, Expr},
    symbol::ParseRestriction,
    Node,
};

#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, thiserror::Error)]
pub enum NodeParseError {
    #[error("unable to parse input")]
    Parse,
    #[error("expected one of {expected}, but received {received}")]
    Expected {
        expected: SyntaxKindSet,
        received: SyntaxKind,
    },
    #[error("call expression array, requires at least a single item, the function to be called.")]
    RequiredFn,
    #[error("malformed call expression array")]
    Array,
    #[error("malformed expression object")]
    Object,
    #[error("expected string to be either a path or signature")]
    PathOrSignature,
}

pub(crate) struct NodeParser<'arena> {
    arena: &'arena Arena,
}

impl<'arena> NodeParser<'arena> {
    pub(crate) const fn new(arena: &'arena Arena) -> Self {
        Self { arena }
    }

    pub(crate) fn parse<'source>(
        &self,
        source: &'source str,
    ) -> Result<Node<'arena, 'source>, NodeParseError> {
        let mut lexer = Lexer::new(source);

        self.parse_node(&mut lexer, None)
    }

    pub(crate) fn parse_node<'source>(
        &self,
        lexer: &mut Lexer<'source>,
        token: Option<Token<'source>>,
    ) -> Result<Node<'arena, 'source>, NodeParseError> {
        let token = if let Some(token) = token {
            token
        } else {
            let mut eof = EofParser { lexer };
            eof.advance().change_context(NodeParseError::Parse)?
        };

        match &token.kind {
            TokenKind::String(..) => self.parse_string(token),
            TokenKind::LBracket => self.parse_call(lexer, token),
            TokenKind::LBrace => self.parse_object(lexer, token),
            _ => Err(Report::new(NodeParseError::Expected {
                expected: SyntaxKindSet::new([
                    SyntaxKind::String,
                    SyntaxKind::LBracket,
                    SyntaxKind::LBrace,
                ]),
                received: SyntaxKind::from(&token.kind),
            })
            .attach(Location::new(token.span))),
        }
    }

    pub(crate) fn parse_call<'source>(
        &self,
        lexer: &mut Lexer<'source>,
        token: Token<'source>,
    ) -> Result<Node<'arena, 'source>, NodeParseError> {
        let mut r#fn = None;
        let mut args = self.arena.vec(None);

        let span = ArrayParser::new(lexer)
            .parse(token, |lexer, token| {
                let node = self.parse_node(lexer, token)?;
                match r#fn {
                    Some(..) => args.push(node),
                    None => r#fn = Some(node),
                }

                Ok(())
            })
            .change_context(NodeParseError::Array)?;

        let Some(r#fn) = r#fn else {
            return Err(Report::new(NodeParseError::RequiredFn).attach(Location::new(span)));
        };

        Ok(Node {
            expr: Expr::Call(Call {
                r#fn: self.arena.boxed(r#fn),
                args: args.into_boxed_slice(),
            }),
            span,
        })
    }

    pub(crate) fn parse_object<'source>(
        &self,
        lexer: &mut Lexer<'source>,
        token: Token<'source>,
    ) -> Result<Node<'arena, 'source>, NodeParseError> {
        NodeObjectParser::new(lexer, self.arena)
            .parse(token)
            .change_context(NodeParseError::Object)
    }

    pub(crate) fn parse_string<'source>(
        &self,
        token: Token<'source>,
    ) -> Result<Node<'arena, 'source>, NodeParseError> {
        let span = token.span;

        let TokenKind::String(value) = token.kind else {
            return Err(Report::new(NodeParseError::Expected {
                expected: SyntaxKindSet::new([SyntaxKind::String]),
                received: SyntaxKind::from(&token.kind),
            })
            .attach(Location::new(span)));
        };

        // TODO: offset error
        let expr = alt((
            parse_signature.map(Expr::Signature),
            parse_path(ParseRestriction::None).map(Expr::Path),
        ))
        .parse(winnow::Stateful {
            input: value.as_ref(),
            state: self.arena,
        })
        .map_err(WinnowError::from)
        .change_context(NodeParseError::PathOrSignature)
        .attach(Location::new(span))?;

        Ok(Node { expr, span })
    }
}

#[cfg(test)]
mod test {
    use insta::assert_debug_snapshot;

    use crate::{arena::Arena, parse::json::node::NodeParser};

    #[test]
    fn fn_is_expr() {
        let arena = Arena::new();

        let result = NodeParser::new(&arena).parse(
            r#"[
            ["input", "variable"],
            "arg1",
            "arg2"
        ]"#,
        );

        assert_debug_snapshot!(result);
    }

    #[test]
    fn fn_empty_args() {
        let arena = Arena::new();

        let result = NodeParser::new(&arena).parse(r#"["func"]"#);

        assert_debug_snapshot!(result);
    }

    #[test]
    fn fn_empty() {
        let arena = Arena::new();

        let result = NodeParser::new(&arena).parse("[]");

        assert_debug_snapshot!(result);
    }

    #[test]
    fn string_is_path() {
        let arena = Arena::new();

        let result = NodeParser::new(&arena).parse(r#""symbol""#);
        assert_debug_snapshot!(result);

        let result = NodeParser::new(&arena).parse(r#""foo::bar""#);
        assert_debug_snapshot!(result);
    }

    #[test]
    fn string_is_signature() {
        let arena = Arena::new();

        let result = NodeParser::new(&arena).parse(r#""<T: Int>(a: T) -> T""#);

        assert_debug_snapshot!(result);
    }

    #[test]
    fn string_is_invalid() {
        let arena = Arena::new();

        let result = NodeParser::new(&arena).parse(r#""1234""#);

        assert_debug_snapshot!(result);
    }

    #[test]
    fn object_is_constant() {
        let arena = Arena::new();

        let result = NodeParser::new(&arena).parse(r#"{"const": 42}"#);

        assert_debug_snapshot!(result);
    }

    #[test]
    fn object_is_constant_with_type() {
        let arena = Arena::new();

        let result = NodeParser::new(&arena).parse(r#"{"type": "u32", "const": 42}"#);

        assert_debug_snapshot!(result);
    }

    #[test]
    fn object_is_constant_with_extra_fields() {
        let arena = Arena::new();

        let result =
            NodeParser::new(&arena).parse(r#"{"type": "u32", "const": 42, "sig": "() -> Unit"}"#);

        assert_debug_snapshot!(result);
    }

    #[test]
    fn object_is_call() {
        let arena = Arena::new();

        let result = NodeParser::new(&arena).parse(r#"{"fn": "func", "args": ["arg1", "arg2"]}"#);

        assert_debug_snapshot!(result);
    }

    #[test]
    fn object_is_args_without_fn() {
        let arena = Arena::new();

        let result = NodeParser::new(&arena).parse(r#"{"args": ["arg1", "arg2"]}"#);

        assert_debug_snapshot!(result);
    }

    #[test]
    fn object_is_call_without_args() {
        let arena = Arena::new();

        let result = NodeParser::new(&arena).parse(r#"{"fn": "func"}"#);

        assert_debug_snapshot!(result);
    }

    #[test]
    fn object_is_signature() {
        let arena = Arena::new();

        let result = NodeParser::new(&arena).parse(r#"{"sig": "<T: Int>(a: T) -> T"}"#);

        assert_debug_snapshot!(result);
    }

    #[test]
    fn object_is_invalid_multiple() {
        let arena = Arena::new();

        let result =
            NodeParser::new(&arena).parse(r#"{"sig": "<T: Int>(a: T) -> T", "fn": "func"}"#);

        assert_debug_snapshot!(result);
    }

    #[test]
    fn object_is_invalid() {
        let arena = Arena::new();

        let result = NodeParser::new(&arena).parse(r#"{"unknown": "key"}"#);

        assert_debug_snapshot!(result);
    }

    #[test]
    fn object_is_empty() {
        let arena = Arena::new();

        let result = NodeParser::new(&arena).parse("{}");

        assert_debug_snapshot!(result);
    }
}
