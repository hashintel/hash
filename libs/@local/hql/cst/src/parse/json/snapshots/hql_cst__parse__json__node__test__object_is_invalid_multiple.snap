---
source: libs/@local/hql/cst/src/parse/json/node.rs
expression: result
---
Err(
    [1mmalformed expression object[22m
    â”œâ•´at [3mlibs/@local/hql/cst/src/parse/json/node.rs:122:14[23m
    â”‚
    â”œâ”€â–¶ [1munable to parse[22m
    â”‚   â•°â•´at [3mlibs/@local/hql/cst/src/parse/json/node_object.rs:514:14[23m
    â”‚
    â”œâ”€â–¶ [1merror while parsing object entry[22m
    â”‚   â•°â•´at [3mlibs/@local/hql/cst/src/parse/json/util.rs:202:55[23m
    â”‚
    â•°â”€â–¶ [1msignature object only allows for `sig` key, received `fn`[22m
        â”œâ•´at [3mlibs/@local/hql/cst/src/parse/json/node_object.rs:431:22[23m
        â•°â•´1 additional opaque attachment,
)
